atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java

```diff
@@ -0,0 +1,214 @@
+                proposer.getQuorumSize(),
+                executor,
+                PaxosQuorumChecker.DEFAULT_REMOTE_REQUESTS_TIMEOUT_IN_SECONDS,
+                true);
+        if (!PaxosQuorumChecker.hasQuorum(responses, proposer.getQuorumSize())) {
+            throw new ServiceNotAvailableException("could not get a quorum");
+        }
+        PaxosLong max = Ordering.natural().onResultOf(PaxosLong::getValue).max(responses);
+        agreedState = getAgreedState(max.getValue());
+        return agreedState.getBound();
+    }
+
+    @VisibleForTesting
+    SequenceAndBound getAgreedState(long seq) {
+        final Optional<SequenceAndBound> state = getLearnedState(seq);
+        if (state.isPresent()) {
+            return state.get();
+        }
+
+        // In the common case seq - 1 will be agreed upon before seq is prepared.
+        Optional<SequenceAndBound> lastState = getLearnedState(seq - 1);
+        if (!lastState.isPresent()) {
+            // We know that even in the case of a truncate, seq - 2 will always be agreed upon.
+            SequenceAndBound forced = forceAgreedState(seq - 2, null);
+            lastState = Optional.of(forceAgreedState(seq - 1, forced.getBound()));
+        }
+
+        return forceAgreedState(seq, lastState.get().getBound());
+    }
+
+    @VisibleForTesting
+    SequenceAndBound forceAgreedState(long seq, @Nullable Long oldState) {
+        if (seq <= PaxosAcceptor.NO_LOG_ENTRY) {
+            return ImmutableSequenceAndBound.of(PaxosAcceptor.NO_LOG_ENTRY, 0L);
+        }
+
+        Optional<SequenceAndBound> state = getLearnedState(seq);
+        if (state.isPresent()) {
+            return state.get();
+        }
+
+        while (true) {
+            try {
+                byte[] value = proposer.propose(seq, oldState == null ? null : PtBytes.toBytes(oldState));
+                // propose must never be null.  We only pass in null for things we know are agreed upon already.
+                Preconditions.checkNotNull(value, "Proposed value can't be null, but was in sequence %s", seq);
+                return ImmutableSequenceAndBound.of(seq, PtBytes.toLong(value));
+            } catch (PaxosRoundFailureException e) {
+                log.info("failed during propose", e);
+                long backoffTime = getRandomBackoffTime();
+                try {
+                    Thread.sleep(backoffTime);
+                } catch (InterruptedException e1) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        }
+    }
+
+    private Optional<SequenceAndBound> getLearnedState(long seq) {
+        if (seq <= PaxosAcceptor.NO_LOG_ENTRY) {
+            return Optional.of(ImmutableSequenceAndBound.of(PaxosAcceptor.NO_LOG_ENTRY, 0L));
+        }
+        List<PaxosLong> responses = PaxosQuorumChecker.collectQuorumResponses(
+                ImmutableList.copyOf(learners),
+                l -> getLearnedValue(seq, l),
+                1,
```

Comment:
Okay. I ported these over without changing them, sorry about that.

`1` here is the minimum quorum size. Since we're talking only to the learners, this is fine, because Paxos guarantees that values that are learned have safely been agreed upon (rather, no 2 learners will learn different values for the same log entry).

The `true` is used by the quorum checker to log failures e.g. `ConnectException`s etc. only if it fails to reach a quorum.

I've renamed these to `QUORUM_OF_ONE` and `ONLY_LOG_ON_QUORUM_FAILURE`, respectively.